### ISSUE TITLE
error retrieving object from ctx

### ISSUE THREAD

**Issue by bkmartinjr** [TileDB-Inc/TileDB-Py#247]
The test program below will fail after repeated creations of context objects.   If you increase or decrease the `sm.num_reader_threads` parameter, the program will fail in more or less iterations, respectively.  If the config param is not set, it will not fail.  My guess is there is a leak in the context code.

```
$ python bug3.py
(0, 64) (0, 58347)
(64, 128) (0, 58347)
(128, 192) (0, 58347)
(192, 256) (0, 58347)
(256, 320) (0, 58347)
(320, 384) (0, 58347)
(384, 448) (0, 58347)
(448, 512) (0, 58347)
(512, 576) (0, 58347)
(576, 640) (0, 58347)
(640, 704) (0, 58347)
(704, 768) (0, 58347)
(768, 832) (0, 58347)
(832, 896) (0, 58347)
(896, 960) (0, 58347)
(960, 1024) (0, 58347)
(1024, 1088) (0, 58347)
(1088, 1152) (0, 58347)
(1152, 1216) (0, 58347)
(1216, 1280) (0, 58347)
(1280, 1344) (0, 58347)
(1344, 1408) (0, 58347)
(1408, 1472) (0, 58347)
(1472, 1536) (0, 58347)
(1536, 1600) (0, 58347)
(1600, 1664) (0, 58347)
(1664, 1728) (0, 58347)
(1728, 1792) (0, 58347)
Traceback (most recent call last):
  File "bug3.py", line 30, in <module>
    main()
  File "bug3.py", line 22, in main
    with tiledb.DenseArray(uri, ctx=tiledb.Ctx(config)) as A:
  File "tiledb/libtiledb.pyx", line 847, in tiledb.libtiledb.Ctx.__init__
  File "tiledb/libtiledb.pyx", line 386, in tiledb.libtiledb._raise_ctx_err
tiledb.libtiledb.TileDBError: error retrieving error object from ctx
```

```
import tiledb

uri = "s3://tiledb-eval/full_hca_uniq_vars-Xr3zstd.cxg/X"

def main():
    config = {
        'sm.num_reader_threads': 128,
    }

    with tiledb.DenseArray(uri) as A:
        nrows, ncols = A.shape

    step = 64
    for row in range(0, nrows, step):
        rows = (row, row + step)
        cols = (0, ncols)
        print(rows, cols)
        with tiledb.DenseArray(uri, ctx=tiledb.Ctx(config)) as A:
            res = A[rows[0]:rows[1], cols[0]:cols[1]]

    return res

if __name__ == '__main__':
    main()
```

**Comment by bkmartinjr** [comment]
Possibly related - I found that I can work around this issue if I use the default context, ie, `tiledb.DenseArray(uri)`, rather than `tiledb.DenseArray(uri, ctx=tiledb.Ctx())`.

This continues to suggest some sort of resource leak or the like.

**Comment by ihnorton** [comment]
Hi @bkmartinjr, yes, the underlying error is that the thread pool allocation failed (at 4096 threads on my macOS computer — system dependent). This is caught internally, but being masked and not returned to tiledb-py. The oversubscription happens because every ctx allocates 128 + (defaults for writer and tbb thread pools. I’m not sure yet if there is a ref count issue with the ctx, or we are running out of threads faster than we can destruct the object and pool. We will definitely improve the error message, and fix any ref count/destructor issues.

Meanwhile, moving the ctx allocation out of the for loop should avoid the issue.

**Comment by bkmartinjr** [comment]
When the Ctx object is destroyed, is the thread pool released?

It sounds like I should shared a context across all threads, rather than allocate them in each thread.   Is that correct?

**Comment by ihnorton** [comment]
> When the Ctx object is destroyed, is the thread pool released?

Yes

> It sounds like I should shared a context across all threads, rather than allocate them in each thread. Is that correct?

Yes, libtiledb is thread-safe, and globally and sharing one Ctx across the whole (python) thread pool is better because schema and fragment metadata is cached per-Ctx.

**Comment by bkmartinjr** [comment]
thanks @ihnorton 

I went down this path as a work-around for some sort of deadlock I ran into while using true ProcessPoolExecutor.  I'll try to isolate that in case it ends up being another bug (more likely user error :-).

**Comment by ihnorton** [comment]
Could you try the following:
```
import multiprocessing as mp
mpx = mp.get_context('spawn')
```
then pass `mp_context=mpx` to `ProcessPoolExecutor`. This will avoid `fork()` for the process creation (but may introduce some overhead in returning the results from each process).

I have located the reference cycle and will be fixing it.

**Comment by bkmartinjr** [comment]
your suggestion worked.  Now that I understand this better, the other alternative is to use the ProcessPoolExecutor `initializer` function to ensure each sub-process creates one and only one Ctx().   I've tried both and they work.  Thank you!

**Comment by ihnorton** [comment]
Fixed by #249

**Comment by ihnorton** [comment]
Thanks again for the bug report. Fix is released in 0.5.3 (ensures Ctx cleanup on destruction, without needing GC run).

Tracking the error improvement message in: https://github.com/TileDB-Inc/TileDB/issues/1457
